xonsh/cli_utils.py

```diff
@@ -51,58 +94,289 @@ def get_doc(func: tp.Callable, parameter: str = None):
     str
         doc of the parameter/function
     """
+    if isinstance(func, str):
+        return func
     import inspect
 
     doc = inspect.getdoc(func) or ""
     if parameter:
-        return _get_param_doc(doc, parameter)
+        return inspect.cleandoc(_get_param_doc(doc, parameter)).strip()
     else:
-        return _get_func_doc(doc)
+        return _get_func_doc(doc).strip()
 
 
 _FUNC_NAME = "_func_"
 
 
+def _get_args_kwargs(annot: tp.Any) -> tp.Tuple[tp.Sequence[str], tp.Dict[str, tp.Any]]:
+    args, kwargs = [], {}
+    if isinstance(annot, tuple):
+        args, kwargs = annot
+    elif TYPING_ANNOTATED_AVAILABLE and "Annotated[" in str(annot):
+        if hasattr(annot, "__metadata__"):
+            args, kwargs = annot.__metadata__[0]
+        else:
+            from typing_extensions import get_args
+
+            _, (args, kwargs) = get_args(annot)
+
+    if isinstance(kwargs, tuple):
+        kwargs = dict(kwargs)
+
+    return args, kwargs
+
+
+def add_args(parser: ap.ArgumentParser, func: tp.Callable, allowed_params=None) -> None:
+    """From Arg -> parser.add_agument(*a, *kw)"""
+    import inspect
+
+    # call this function when this sub-command is selected
+    parser.set_defaults(**{_FUNC_NAME: func})
+
+    sign = inspect.signature(func)
+    for name, param in sign.parameters.items():
+        if name.startswith("_") or (
+            allowed_params is not None and name not in allowed_params
+        ):
+            continue
+        args, kwargs = _get_args_kwargs(param.annotation)
+
+        if args:  # optional argument. eg. --option
+            kwargs.setdefault("dest", name)
+        else:  # positional argument
+            args = [name]
+
+        if inspect.Parameter.empty != param.default:
+            kwargs.setdefault("default", param.default)
+
+        kwargs.setdefault("help", get_doc(func, name))
+        completer = kwargs.pop("completer", None)
+        action = parser.add_argument(*args, **kwargs)
+        if completer:
+            action.completer = completer  # type: ignore
+        action.help = action.help or ""
+        if action.default and "%(default)s" not in action.help:
+            action.help += os.linesep + " (default: %(default)s)"
+        if action.type and "%(type)s" not in action.help:
+            action.help += " (type: %(type)s)"
```

Comment:
After all of this we should set `action.description = action.help.split("\n", 1)[0]` and use it later in `RichCompletion.description` for two reasons:
1. The `%(default)` and `%(type)` don't get formatted in the completion's description
2. The newline doesn't render well as the completion's description

Reply:
they are displayed in --help output though. In future, I want to have a completer that can show multi-line description, much like IDEs.

Reply:
That's why I said - keep the help string as is for `--help` and only change the description for RichCompletion.
The multi-line description thing could be cool and can be in a separate pr when you get to it, but for now it doesn't work so let's keep the first line only

Reply:
here this is help output. we don't need to truncate here I suppose.

Reply:
Instead of using the same text for both `--help` and `RichCompletion.description` , create a shortened version for the description without these `(default: ...)` and `(type: ...)` fields and only the first line (until newlines can be rendered in ptk)

Reply:
The completer should not be making assumptions about the interface. What if I want to use fzf for completions while ptk completion is disabled? It supports multiple lines.

Reply:
Ok that makes sense.
So we need to fill in the `type` and `default` fields when setting `RichCompletion.description`.
Also maybe replace the newlines with a period and a space in `xonsh/ptk` for the time being so we won't have the ugly `^J`?

Reply:
still replacing \n should be done inside ptk-completer as it is interface specific. I am not sure about the short representation. don't we want to see if there is a default and what type of argument expected. if there is space to show that info, don't we want to see ?

Reply:
created PR for this https://github.com/xonsh/xonsh/pull/4419

Reply:
> don't we want to see if there is a default and what type of argument expected. if there is space to show that info, don't we want to see ?

Yeah I agree! Just last time I checked the type and default didn't work in the description

Reply:
![image](https://user-images.githubusercontent.com/6702219/130031973-c02aed3a-5833-41a6-930e-9e1796e9331a.png)

are you talking about --help ?

Reply:
I saw this happen before you split the PRs:
If I change `pos` to `--pos` and add a `%(default)` to the description it doesn't render the value:
![image](https://user-images.githubusercontent.com/18242949/130503905-8f435fa0-0d26-4cde-b640-be5b50e84cf0.png)

Reply:
with this https://github.com/xonsh/xonsh/pull/4391/commits/f91e204545892e23fc0f7da18873bc3b89170341 above will get expanded if available.